extensions/amp-story/1.0/amp-story.js

```diff
@@ -2772,6 +2804,43 @@ export class AmpStory extends AMP.BaseElement {
         devError(TAG, err, 'Bundle not found for language ' + languageCode);
       });
   }
+
+  /**
+   * @return {number} The index of the last page that should be shown in
+   *     preview mode.
+   * @private
+   */
+  calculateIndexOfLastPageToPreview_() {
+    const minPreviewPages =
+      parseInt(this.viewer_?.getParam('minPreviewPages'), 10) ??
+      DEFAULT_MIN_PAGES_TO_PREVIEW;
+    const pctPagesToPreview =
+      parseInt(this.viewer_?.getParam('pctPagesToPreview'), 10) ??
+      DEFAULT_PCT_PAGES_TO_PREVIEW;
+
+    // We calculate the number of preview pages by taking the larger of the two
+    // values: min # of preview pages vs the % of pages to show.
+    const numPages = this.element.querySelectorAll('amp-story-page').length;
+    let numPreviewPages = Math.ceil((pctPagesToPreview / 100) * numPages);
+    numPreviewPages = Math.max(minPreviewPages, numPreviewPages);
+    numPreviewPages = clamp(numPreviewPages, 1, numPages);
+
+    return numPreviewPages - 1;
+  }
+
+  /**
+   * Sends the 'storyPreviewFinished' viewer message if the active page is
+   * supposed to be the last page shown in the preview.
+   * @private
+   */
+  sendMessageIfPreviewFinished_() {
+    const pageId = this.activePage_.element.id;
+    const lastPreviewPage = this.pages_[this.indexOfLastPageToPreview_];
+    const isLastPreviewPage = pageId === lastPreviewPage.element.id;
```

Comment:
A better logic here could be (will write pseudocode) `const index = this.pages.indexOf(activePage); const finishedPreviewPages = index === this.indexOfLastPageToPreview_` so that this works well with things like branching or opening stories from non-first pages.